File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        output = []
        
        def backtrack(nums, idx, inner):
            output.append(inner.copy())
            
            for i in range(idx, len(nums)):
                backtrack(nums, i+1, inner + [nums[i]])
            
        backtrack(nums, 0, [])
        return output
        

File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList();
        backtrack(nums, 0, new ArrayList<Integer>(), result);
        return result;
    }
    
    public void backtrack(int[] nums, int idx, List<Integer> inter, List<List<Integer>> result){
        result.add(new ArrayList<Integer>(inter));
        
        for(int i=idx; i<nums.length; i++){
            inter.add(nums[i]);
            backtrack(nums, i+1, inter, result);
            inter.remove(inter.size()-1);
        }
    }
}